x-pack/plugins/lens/public/xy_visualization_plugin/xy_expression.tsx

```diff
@@ -0,0 +1,268 @@
+  type: 'lens_xy_legendConfig',
+  help: `Configure the xy chart's legend`,
+  context: {
+    types: ['null'],
+  },
+  args: {
+    isVisible: {
+      types: ['boolean'],
+      help: 'Specifies whether or not the legend is visible.',
+    },
+    position: {
+      types: ['string'],
+      options: [Position.Top, Position.Right, Position.Bottom, Position.Left],
+      help: 'Specifies the legend position.',
+    },
+  },
+  fn: function fn(_context: unknown, args: LegendConfig) {
+    return {
+      type: 'lens_xy_legendConfig',
+      ...args,
+    };
+  },
+};
+
+interface AxisConfig {
+  title: string;
+  showGridlines: boolean;
+  position: Position;
+}
+
+const axisConfig: { [key in keyof AxisConfig]: ArgumentType<AxisConfig[key]> } = {
+  title: {
+    types: ['string'],
+    help: 'The axis title',
+  },
+  showGridlines: {
+    types: ['boolean'],
+    help: 'Show / hide axis grid lines.',
+  },
+  position: {
+    types: ['string'],
+    options: [Position.Top, Position.Right, Position.Bottom, Position.Left],
+    help: 'The position of the axis',
+  },
+};
+
+export interface YConfig extends AxisConfig {
+  accessors: string[];
+}
+
+type YConfigResult = YConfig & { type: 'lens_xy_yConfig' };
+
+export const yConfig: ContextFunction<'lens_xy_yConfig', null, YConfig, YConfigResult> = {
+  name: 'lens_xy_yConfig',
+  aliases: [],
+  type: 'lens_xy_yConfig',
+  help: `Configure the xy chart's y axis`,
+  context: {
+    types: ['null'],
+  },
+  args: {
+    ...axisConfig,
+    accessors: {
+      types: ['string'],
+      help: 'The columns to display on the y axis.',
+      multi: true,
+    },
+  },
+  fn: function fn(_context: unknown, args: YConfig) {
+    return {
+      type: 'lens_xy_yConfig',
+      ...args,
+    };
+  },
+};
+
+export interface XConfig extends AxisConfig {
+  accessor: string;
+}
+
+type XConfigResult = XConfig & { type: 'lens_xy_xConfig' };
+
+export const xConfig: ContextFunction<'lens_xy_xConfig', null, XConfig, XConfigResult> = {
+  name: 'lens_xy_xConfig',
+  aliases: [],
+  type: 'lens_xy_xConfig',
+  help: `Configure the xy chart's x axis`,
+  context: {
+    types: ['null'],
+  },
+  args: {
+    ...axisConfig,
+    accessor: {
+      types: ['string'],
+      help: 'The column to display on the x axis.',
+    },
+  },
+  fn: function fn(_context: unknown, args: XConfig) {
+    return {
+      type: 'lens_xy_xConfig',
+      ...args,
+    };
+  },
+};
+
+export interface XYArgs {
+  seriesType: 'bar' | 'line' | 'area';
+  title: string;
+  legend: LegendConfigResult;
+  y: YConfigResult;
+  x: XConfigResult;
+  splitSeriesAccessors: string[];
+  stackAccessors: string[];
+}
+
+export interface XYChartProps {
+  data: KibanaDatatable;
+  args: XYArgs;
+}
+
+// TODO: Specify the TypeScript type of this definition, once the
+// ContextFunction has moved to core and has the correct signature:
+// ContextFunction<'lens_xy_chart', KibanaDatatable, XYArgs, XYRender>
+export const xyChart = {
```

Comment:
As you are already using ContextFunction in line 129, do you want to just use the canvas version here or did you plan to add a similar TODO the the xConfig function definition?

Reply:
It won't work here, because the ContextFunction doesn't support custom complex types in the args. All of the other places I used it (e.g. line 129) are using primitive types in their args.

Reply:
Bummer